### PR TITLE
[REVIEW] Add support to check for "NaT" and "None" strings while typecasting to `datetime64`

### PR DESCRIPTION
**Pull request by galipremsagar**
Fixes: #5117 

1. Checking for `'None'` values and erroring to match pandas behavior.
2. Setting the result values to `null` if we find any `'NaT'` values in the input column.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/cudf/cudf/core/column/string.py [hunk 1]
```diff
@@ -2053,34 +2053,40 @@ def _nbytes(self):
...
                     "Could not convert strings to float \
                         type due to presence of non-floating values."
                 )
 
-        return _str_to_numeric_typecast_functions[str_dtype](self, **kwargs)
+        result_col = _str_to_numeric_typecast_functions[out_dtype](
+            self, **kwargs
+        )
+        if (out_dtype.type is np.datetime64) and boolean_match.any():
+            result_col[boolean_match] = None
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/string.py:2088]: `NaT` values are the datetime equivalent of `NaN` values where we should be able to control if they get converted to `null` or stay as `NaT` if we support them. That being said, `NaT` is a numpy / Pandas specific thing.

For example, Arrow doesn't have a concept of `NaT`. I would be in strong support of not supporting `NaT` other than in going from string --> datetime64 and then converting it to null always.

Any thoughts?

File: python/cudf/cudf/core/column/string.py [hunk 2]
```diff
@@ -2053,34 +2053,40 @@ def _nbytes(self):
...
                 if len(self) > 0:
                     # infer on host from the first not na element
                     fmt = datetime.infer_format(self[self.notna()][0])
                     kwargs.update(format=fmt)
-        kwargs.update(dtype=out_dtype)
 
-        if str_dtype.kind in ("i"):
+            # Check for None strings
+            if len(self) > 0 and self.binary_operator("eq", "None").any():
+                raise ValueError("Could not convert `None` value to datetime")
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/core/column/string.py:2068]: Pandas allow `None` in `object` but not `'None'` in `object` while converting string --> datetime64

```python
>>> import pandas as pd
>>> pd.Series(["2001-01-01", "2002-02-02", "2000-01-05", "NaT"]).astype('datetime64[s]')
0   2001-01-01
1   2002-02-02
2   2000-01-05
3          NaT
dtype: datetime64[ns]
>>> pd.Series(["2001-01-01", "2002-02-02", "2000-01-05", "None"]).astype('datetime64[s]')
Traceback (most recent call last):
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/arrays/datetimes.py", line 1979, in objects_to_datetime64ns
    values, tz_parsed = conversion.datetime_to_datetime64(data)
  File "pandas/_libs/tslibs/conversion.pyx", line 200, in pandas._libs.tslibs.conversion.datetime_to_datetime64
TypeError: Unrecognized value type: <class 'str'>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/generic.py", line 5882, in astype
    dtype=dtype, copy=copy, errors=errors, **kwargs
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/internals/managers.py", line 581, in astype
    return self.apply("astype", dtype=dtype, **kwargs)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/internals/managers.py", line 438, in apply
    applied = getattr(b, f)(**kwargs)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/internals/blocks.py", line 559, in astype
    return self._astype(dtype, copy=copy, errors=errors, values=values, **kwargs)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/internals/blocks.py", line 643, in _astype
    values = astype_nansafe(vals1d, dtype, copy=True, **kwargs)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/dtypes/cast.py", line 715, in astype_nansafe
    return astype_nansafe(to_datetime(arr).values, dtype, copy=copy)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/util/_decorators.py", line 208, in wrapper
    return func(*args, **kwargs)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/tools/datetimes.py", line 794, in to_datetime
    result = convert_listlike(arg, box, format)
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/tools/datetimes.py", line 463, in _convert_listlike_datetimes
    allow_object=True,
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/arrays/datetimes.py", line 1984, in objects_to_datetime64ns
    raise e
  File "/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/arrays/datetimes.py", line 1975, in objects_to_datetime64ns
    require_iso8601=require_iso8601,
  File "pandas/_libs/tslib.pyx", line 465, in pandas._libs.tslib.array_to_datetime
  File "pandas/_libs/tslib.pyx", line 688, in pandas._libs.tslib.array_to_datetime
  File "pandas/_libs/tslib.pyx", line 822, in pandas._libs.tslib.array_to_datetime_object
  File "pandas/_libs/tslib.pyx", line 813, in pandas._libs.tslib.array_to_datetime_object
  File "pandas/_libs/tslibs/parsing.pyx", line 225, in pandas._libs.tslibs.parsing.parse_datetime_string
  File "/conda/envs/cudf/lib/python3.7/site-packages/dateutil/parser/_parser.py", line 1374, in parse
    return DEFAULTPARSER.parse(timestr, **kwargs)
  File "/conda/envs/cudf/lib/python3.7/site-packages/dateutil/parser/_parser.py", line 649, in parse
    raise ParserError("Unknown string format: %s", timestr)
dateutil.parser._parser.ParserError: Unknown string format: None
>>> pd.Series(["2001-01-01", "2002-02-02", "2000-01-05", None]).astype('datetime64[s]')
0   2001-01-01
1   2002-02-02
2   2000-01-05
3          NaT
dtype: datetime64[ns]
>>> import cudf
>>> cudf.Series(["2001-01-01", "2002-02-02", "2000-01-05", None]).astype('datetime64[s]')
0   2001-01-01
1   2002-02-02
2   2000-01-05
3         null
dtype: datetime64[s]
>>> cudf.Series(["2001-01-01", "2002-02-02", "2000-01-05", "NaT"]).astype('datetime64[s]')
0   2001-01-01
1   2002-02-02
2   2000-01-05
3         null
dtype: datetime64[s]
>>> cudf.Series(["2001-01-01", "2002-02-02", "2000-01-05", "None"]).astype('datetime64[s]')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/conda/envs/cudf/lib/python3.7/site-packages/cudf/core/series.py", line 1473, in astype
    raise e
  File "/conda/envs/cudf/lib/python3.7/site-packages/cudf/core/series.py", line 1465, in astype
    data = self._column.astype(dtype, **kwargs)
  File "/conda/envs/cudf/lib/python3.7/site-packages/cudf/core/column/column.py", line 824, in astype
    return self.as_datetime_column(dtype, **kwargs)
  File "/conda/envs/cudf/lib/python3.7/site-packages/cudf/core/column/string.py", line 2092, in as_datetime_column
    return self.as_numerical_column(dtype, **kwargs)
  File "/conda/envs/cudf/lib/python3.7/site-packages/cudf/core/column/string.py", line 2068, in as_numerical_column
    raise ValueError("Could not convert `None` value to datetime")
ValueError: Could not convert `None` value to datetime
```

File: python/cudf/cudf/core/column/string.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/column/string.py:2088]: > `That being said, `NaT` is a numpy / Pandas specific thing.

+1

> 
> For example, Arrow doesn't have a concept of `NaT`. I would be in strong support of not supporting `NaT` other than in going from string --> datetime64 and then converting it to null always.
> 
> Any thoughts?

Agree on this, I think we should stick to Arrow behavior except handling "NaT" in string to datetime64 conversions, but I don't see any other flows in which "NaT" can be introduced?
For example we already handle `NaT` as `null` when we convert a pandas object to cudf.

```python
>>> cudf.from_pandas(pd.Series(["2001-01-01", "2002-02-02", "2000-01-05", "NaT"]).astype('datetime64[s]'))
0   2001-01-01
1   2002-02-02
2   2000-01-05
3         null
dtype: datetime64[ns]
```

**Comment by galipremsagar** [comment]
rerun tests

**Pull request merged**